test/virtual/host.cpp

```diff
@@ -11,13 +11,143 @@
+
 extern "C"
 {
+    const char* _ecall_table[OE_MAX_ECALLS];
+    uint32_t _ecall_table_size;
     thread_local oe_enclave_t* _enclave;
 
+    // Get the global ecall id from the _ecall_table.
+    static oe_result_t oe_get_global_ecall_id_by_name(
+        const char* name,
+        uint64_t* global_id)
+    {
+        oe_result_t result = OE_NOT_FOUND;
+        uint64_t i;
+
+        if (!global_id)
+        {
+            result = OE_INVALID_PARAMETER;
+            goto done;
+        }
+
+        for (i = 0; i < _ecall_table_size; i++)
+        {
+            if (strcmp(_ecall_table[i], name) == 0)
+            {
+                *global_id = i;
+                result = OE_OK;
+                break;
+            }
+        }
+
+        if (result == OE_NOT_FOUND)
+        {
+            if (_ecall_table_size >= OE_MAX_ECALLS)
+            {
+                result = OE_OUT_OF_BOUNDS;
+                goto done;
+            }
+
+            _ecall_table[_ecall_table_size] = name;
+            *global_id = _ecall_table_size;
+            _ecall_table_size++;
+            result = OE_OK;
+        }
+
+    done:
+        return result;
+    }
+
+    static oe_result_t oe_get_enclave_function_id(
+        oe_enclave_t* enclave,
+        uint64_t* global_id,
+        const char* name,
+        uint64_t* id)
+    {
+        oe_result_t result = OE_FAILURE;
+
+        if (!enclave || !global_id || !name || !id)
+        {
+            result = OE_INVALID_PARAMETER;
+            goto done;
+        }
+
+        if (*global_id == OE_GLOBAL_ECALL_ID_NULL)
+        {
+            if ((result = oe_get_global_ecall_id_by_name(name, global_id)) !=
+                OE_OK)
+                goto done;
+        }
+
+        if (*global_id >= OE_MAX_ECALLS)
+        {
+            result = OE_OUT_OF_BOUNDS;
+            goto done;
+        }
+
+        /* Look-up the ecall id from the per-enclave table. */
+        *id = enclave->_ecall_info_table[*global_id].id;
+
+        result = OE_OK;
+    done:
+
+        return result;
+    }
+
+    static oe_result_t oe_host_register_enclave_functions(
+        oe_enclave_t* enclave,
+        const char** ecall_name_list,
+        uint32_t num_ecalls)
+    {
+        oe_result_t result = OE_UNEXPECTED;
+
+        if (!enclave || num_ecalls != enclave->_host_num_ecalls)
+        {
+            result = OE_INVALID_PARAMETER;
+            goto done;
+        }
+
+        /*  Nothing to add when the table is empty, fall through. */
+        if (!ecall_name_list || !num_ecalls)
+        {
+            result = OE_OK;
+            goto done;
+        }
+
+        uint32_t i;
+        for (i = 0; i < num_ecalls; i++)
+        {
+            uint64_t global_id = OE_GLOBAL_ECALL_ID_NULL;
+            const char* name = ecall_name_list[i];
+            uint64_t local_id = i;
+
+            /* Avoid registering dubplicated functions. */
+            if ((oe_get_global_ecall_id_by_name(name, &global_id)) != OE_OK)
```

Comment:
Do we need this check? I'm fine with the check; but am curious if you were thinking of a specific scenario.